# REDES NEURONALES - REGRESIÓN LOGÍSTICA MULTICLASE 

### Nombre: Gonzales Suyo Franz Reinaldo

### Carrera: Ing. de Sistemas

### C.U. 35-5335

# Implementación del modelo de regresión logística multiclas - Redes Neuronales

En este ejercicio se implementa crear una red neuronal para la regresion logistica multiclase y se aplica a dos diferentes datasets.

Nuestro objetuvo es predecir el estado de un hombre de estado civil, donde nuestro objetivo es predecir si el adulto tienen varios estados como: casado, soltero, viudo, sin-familia, etc.
En el siguiente dataset `adult.data.csv` se encuentran todos los datos sobre el estado del adulto.
Link del Dataset: https://data.world/uci/adult

Enlace del GitHub: 

# **1: Preprocesamiento de Datos**

# Informacióndel Dataset

Especificar cuales son sus  variables de entrada  (X): 

Variables de entrada son: 

X1:age: continuous.

X2:workclass: Private, Self-emp-not-inc, Self-emp-inc, Federal-gov, Local-gov, State-gov, Without-pay, Never-worked.

X3: fnlwgt: continuous.

X4: education: Bachelors, Some-college, 11th, HS-grad, Prof-school, Assoc-acdm, Assoc-voc, 9th, 7th-8th, 12th, Masters, 1st-4th, 10th, Doctorate, 5th-6th, Preschool.

X5: education-num: continuous.

X6: marital-status: Married-civ-spouse, Divorced, Never-married, Separated, Widowed, Married-spouse-absent, Married-AF-spouse.

X7: occupation: Tech-support, Craft-repair, Other-service, Sales, Exec-managerial, Prof-specialty, Handlers-cleaners, Machine-op-inspct, Adm-clerical, Farming-fishing, Transport-moving, Priv-house-serv, Protective-serv, Armed-Forces.

X8: relationship: Wife, Own-child, Husband, Not-in-family, Other-relative, Unmarried.

X9: race: White, Asian-Pac-Islander, Amer-Indian-Eskimo, Other, Black.

X10: sex: Female, Male.

X11: capital-gain: continuous.

X12: capital-loss: continuous.

X13: hours-per-week: continuous.

X14: native-country: United-States, Cambodia, England, Puerto-Rico, Canada, Germany, Outlying-US(Guam-USVI-etc), India, Japan, Greece, South, China, Cuba, Iran, Honduras, Philippines, Italy, Poland, Jamaica, Vietnam, Mexico, Portugal, Ireland, France, Dominican-Republic, Laos, Ecuador, Taiwan, Haiti, Columbia, Hungary, Guatemala, Nicaragua, Scotland, Thailand, Yugoslavia, El-Salvador, Trinadad&Tobago, Peru, Hong, Holand-Netherlands.

X15: class: >50K, <=50K

Y Label(Predicha):  Columna X6 "marital-status"



Este dataset tiene una conjunto de datos donde reconoce el tipo de estado civil de un adulto que puede estar
en diferentes tipos de estados o clases como: Married-civ-spouse, Divorced, Never-married, Separated, Widowed, Married-spouse-absent, Married-AF-spouse.

Se inicia el ejercicio cargando los datos 

## Importación de las Librerias

In [1]:

# Plotting library
from matplotlib import pyplot

# se utiliza para el manejo de rutas y directorios.
import os

# Calculo cientifico y vectorial para python
import numpy as np

# Librerias para graficar
import matplotlib.pyplot as plt

import pandas as pd

# Modulo de optimización de scipy
from scipy import optimize

#Para separa el Dataset 20% y 80% para diferentes pruebas
from sklearn.model_selection import train_test_split

# para aumentar datos en un dataset
from collections import Counter
from imblearn.over_sampling import SMOTE

# le dice a matplotlib que incruste gráficos en el cuaderno
%matplotlib inline

# tells matplotlib to embed plots within the notebook
%matplotlib inline

## 1.1 Obtención y Preparación de Datos:

In [2]:
# Obtenemos los datos 
df_train = pd.read_csv('./adult_train_data.csv')
df_test = pd.read_csv('./adult_test_data.csv')

# Datos de entrenamiento del 80% 
X_train = df_train.drop('marital-status', axis=1)
y_train = df_train['marital-status']


# Datos del prueba del 20 % 
X_test = df_test.drop('marital-status', axis=1)
y_test = df_test['marital-status']

# Datos de X_train para el entrenamiento 
print("Datos de X_train")
print(X_train)

print('=' * 100)

print("Datos de y_train")
print(y_train)

# Mostramos la cantidad de ejemplos que se utilizaran para el entrenamiento
print('=' * 100)
print("Cantidad de ejemplos del 80% para el entrenamiento es de: {:.0f}".format(len(X_train)))
print("Cantidad de ejemplos del 20% para la prueba es de: {:.0f}".format(len(X_test)))

Datos de X_train
       age  workclass  fnlwgt  education  education-num  occupation  \
0       20          4  170800         15             10           5   
1       39          4  111268          9             13           1   
2       17          4  347322          0              6          12   
3       24          4  103064          9             13          10   
4       46          4  376789         11              9           8   
...    ...        ...     ...        ...            ...         ...   
25689   24          4  185821         15             10          12   
25690   26          4  193945          7             12          13   
25691   44          4   79531          8             11           3   
25692   25          4   28473          8             11           3   
25693   36          4  139743         15             10          10   

       relationship  race  sex  capital-gain  capital-loss  hours-per-week  \
0                 3     4    0             0        

In [3]:
# Configurar Pandas para que no corte la visualización
pd.set_option('display.max_rows', 50)  # Mostrar todas las filas (60 -> None)
pd.set_option('display.max_columns', None)  # Mostrar todas las columnas (20 -> None)

In [4]:
# Mostramos cuantas clases tinene la columna de "Y" marital-status
df_train['marital-status'].value_counts()

marital-status
0    11980
2     8546
1     3554
3      820
4      794
Name: count, dtype: int64

In [5]:
y_train = np.array([int(e) for e in y_train])
print(y_train.shape)
y_train = np.squeeze(y_train)

print(X_train.shape)
print(y_train.shape)
print(X_test.shape)
print(y_test.shape)

(25694,)
(25694, 14)
(25694,)
(6422, 14)
(6422,)


In [6]:
# Mostramos los datos de entrenamiento X_train:
X_train

age  workclass  fnlwgt  education  education-num  occupation  \
0       20          4  170800         15             10           5   
1       39          4  111268          9             13           1   
2       17          4  347322          0              6          12   
3       24          4  103064          9             13          10   
4       46          4  376789         11              9           8   
...    ...        ...     ...        ...            ...         ...   
25689   24          4  185821         15             10          12   
25690   26          4  193945          7             12          13   
25691   44          4   79531          8             11           3   
25692   25          4   28473          8             11           3   
25693   36          4  139743         15             10          10   

       relationship  race  sex  capital-gain  capital-loss  hours-per-week  \
0                 3     4    0             0             0              40   
1                 3     4    0             0             0              40   
2                 3     4    0             0             0              20   
3                 1     4    0             0             0              40   
4                 1     4    1             0             0              15   
...             ...   ...  ...           ...           ...             ...   
25689             3     4    0             0             0              40   
25690             1     4    1             0             0              45   
25691             0     4    1             0             0              40   
25692             0     4    1             0             0              40   
25693             3     4    0             0             0              40   

       native-country  class  
0                  39      0  
1                  39      0  
2                  39      0  
3                  39      0  
4                  39      0  
...               ...    ...  
25689              39      0  
25690              39      0  
25691              39      1  
25692              39      0  
25693              39      0  

[25694 rows x 14 columns]

## **2. Normalización de Datos** 

In [7]:
def  featureNormalize(X):
    X_norm = X.copy()
    mu = np.zeros(X.shape[1])
    sigma = np.zeros(X.shape[1])

    mu = np.mean(X, axis = 0)
    sigma = np.std(X, axis = 0)
    X_norm = (X - mu) / sigma

    return X_norm, mu, sigma

In [8]:
# Llamamos a la function de featureNormalize para normalizar los datos de entrenamiento
X_norm, mu, sigma = featureNormalize(X_train)

# Mostramos los datos de X_train normalizados
X_norm

age  workclass    fnlwgt  education  education-num  occupation  \
0     -1.358529   0.088814 -0.179037   1.214815      -0.036478   -0.373208   
1      0.029830   0.088814 -0.746179  -0.338802       1.139008   -1.317190   
2     -1.577744   0.088814  1.502630  -2.669226      -1.603792    1.278761   
3     -1.066243   0.088814 -0.824336  -0.338802       1.139008    0.806770   
4      0.541330   0.088814  1.783353   0.179070      -0.428306    0.334779   
...         ...        ...       ...        ...            ...         ...   
25689 -1.066243   0.088814 -0.035937   1.214815      -0.036478    1.278761   
25690 -0.920100   0.088814  0.041458  -0.856674       0.747180    1.514756   
25691  0.395187   0.088814 -1.048528  -0.597738       0.355351   -0.845199   
25692 -0.993171   0.088814 -1.534941  -0.597738       0.355351   -0.845199   
25693 -0.189385   0.088814 -0.474907   1.214815      -0.036478    0.806770   

       relationship      race       sex  capital-gain  capital-loss  \
0          0.974253  0.388385 -1.432448     -0.146356     -0.216102   
1          0.974253  0.388385 -1.432448     -0.146356     -0.216102   
2          0.974253  0.388385 -1.432448     -0.146356     -0.216102   
3         -0.270524  0.388385 -1.432448     -0.146356     -0.216102   
4         -0.270524  0.388385  0.698106     -0.146356     -0.216102   
...             ...       ...       ...           ...           ...   
25689      0.974253  0.388385 -1.432448     -0.146356     -0.216102   
25690     -0.270524  0.388385  0.698106     -0.146356     -0.216102   
25691     -0.892912  0.388385  0.698106     -0.146356     -0.216102   
25692     -0.892912  0.388385  0.698106     -0.146356     -0.216102   
25693      0.974253  0.388385 -1.432448     -0.146356     -0.216102   

       hours-per-week  native-country     class  
0           -0.034413        0.283824 -0.565754  
1           -0.034413        0.283824 -0.565754  
2           -1.661160        0.283824 -0.565754  
3           -0.034413        0.283824 -0.565754  
4           -2.067847        0.283824 -0.565754  
...               ...             ...       ...  
25689       -0.034413        0.283824 -0.565754  
25690        0.372273        0.283824 -0.565754  
25691       -0.034413        0.283824  1.767551  
25692       -0.034413        0.283824 -0.565754  
25693       -0.034413        0.283824 -0.565754  

[25694 rows x 14 columns]

In [9]:
# Imprimos todas las clases que contiene la columna de y_train
from collections import Counter

num_clases = len(np.unique(y_train))
print("Número de clases:", num_clases)


# Conteo de datos por clase
conteo_clases = dict(Counter(y_train))

# Imprimir el conteo de datos por clase
print("Clases podrían ser:")
for clase, conteo in conteo_clases.items():
    print(f"{clase} : {conteo} datos")

Número de clases: 5
Clases podrían ser:
2 : 8546 datos
0 : 11980 datos
3 : 820 datos
4 : 794 datos
1 : 3554 datos


## **3. Construcción del Modelo de Red Neuronal**

### 3.1 Definición de Capas

In [10]:
# Contrucción del modelo de red neuronal
# Configuración de parámetros necesario
input_layer_size  = 14  # 14 features, caracteristicas
hidden_layer_size = 10   # 10 hidden units, neuronas ocultas, capa uculta
num_labels = 5         # 5 labels, clases, de 0 a 4


# carga los pesos en las variables Theta1 y Theta2
pesos = {}
pesos['Theta1'] = np.random.rand(hidden_layer_size, input_layer_size + 1) # 10, 15
pesos['Theta2'] = np.random.rand(num_labels, hidden_layer_size + 1)  # 5, 11

Theta1 = pesos['Theta1']
Theta2 = pesos['Theta2']

# Desenrollar parámetros
print(Theta1.ravel().shape)  # ravel() para convertir los pesos de una matriz a un vector
print(Theta2.ravel().shape)

nn_params = np.concatenate([Theta1.ravel(), Theta2.ravel()])
print(nn_params.shape)

(150,)
(55,)
(205,)


### 3.2 Función de Activación

La principal diferencia es que, en el modelo de `regresión logísitca`, utilizaremos una función de activación conocida como `Sigmoid`.

$$ \sigma(z) = \frac{1}{1 + e^{-z}} $$

In [11]:
# Función de activación 
def sigmoid(z):
    """
    Computes the sigmoid of z.
    """
    return 1.0 / (1.0 + np.exp(-z))


def sigmoidGradient(z):

    g = np.zeros(z.shape)

    g = sigmoid(z) * (1 - sigmoid(z))

    return g

### 3.3 Función de Pérdida (Costo) y Descenso del Gradiente

Podríamos intentar entrenar nuestro modelo de `regresión logísitca multiclase` con la función de pérdida que ya conocemos. Sin embargo. Esta función es conocida como *Cross Entropy*.

$$ J(\mathbf{w}) = - \frac{1}{N} \sum^{N}_{j=1} \left[ y^{(j)} log(\hat{y}^{(j)}) + (1-y^{(j)})log(1-\hat{y}^{(j)})\right] $$


In [12]:
# Definimos la función 
# Esta función calcula la función de costo y los gradientes para una red neuronal de dos capas (una capa oculta) 
# utilizada en un problema de regresión logística multiclase.

def nnCostFunction(nn_params, input_layer_size, hidden_layer_size, num_labels, X, y, lambda_):
    
    # Reformar nn_params nuevamente en los parámetros Theta1 y Theta2, las matrices de peso
    # para nuestra red neuronal de 2 capas
    
    #? Los primeros 150 elementos de nn_params se utilizan para llenar la matriz Theta1, 
    # que tendrá dimensiones 10 x 15 (10 filas y 15 columnas).
    Theta1 = np.reshape(nn_params[:hidden_layer_size * (input_layer_size + 1)],
                        (hidden_layer_size, (input_layer_size + 1)))
    
    #? Los elementos apartir de 155 de nn_params se utilizan para llenar la matriz Theta2, 
    # entonces 55 elementos que tendrá dimensiones 5 x 11 (5 filas y 11 columnas.
    Theta2 = np.reshape(nn_params[(hidden_layer_size * (input_layer_size + 1)):],
                        (num_labels, (hidden_layer_size + 1)))
    
    m = y.size
    
    J = 0
    
    #? Inicializamos matrices para almacenar los gradientes
    Theta2_grad = np.zeros(Theta1.shape)
    Theta2_grad = np.zeros(Theta2.shape)
    
    a1 = np.concatenate([np.ones((m, 1)), X], axis=1)
    
    a2 = sigmoid(a1.dot(Theta1.T))
    a2 = np.concatenate([np.ones((a2.shape[0], 1)), a2], axis=1)
    
    a3 = sigmoid(a2.dot(Theta2.T))
    
    y_matrix = y.reshape(-1)
    y_matrix = np.eye(num_labels)[y_matrix]
    
    temp1 = Theta1
    temp2 = Theta2
    
    
    #? Calculo de la función de costo con termino de regularización
    
    regularizacion = (lambda_ / (2 * m)) * (np.sum(np.square(temp1[:, 1:])) + np.sum(np.square(temp2[:, 1:])))
    
    J = (-1 / m) * np.sum((y_matrix * np.log(a3)) + (1 - y_matrix) * np.log(1 - a3)) + regularizacion
    
    
    #? Backpropagation
    
    # calcula el error en la capa de salida.
    delta_3 = a3 - y_matrix 
    # calcula el error en la capa oculta propagando hacia atrás el error desde la capa de salida.
    delta_2 = delta_3.dot(Theta2)[:, 1:] * sigmoidGradient(a1.dot(Theta1.T))
    
    # Acumuladores de los gradientes de los pesos Theta1 y Theta2 respectivamente.
    Delta1 = delta_2.T.dot(a1)
    Delta2 = delta_3.T.dot(a2)
    
    # Agregar regularización al gradiente
    
    Theta1_grad = (1 / m) * Delta1
    Theta1_grad[:, 1:] = Theta1_grad[:, 1:] + (lambda_ / m) * Theta1[:, 1:]
    
    Theta2_grad = (1 / m) * Delta2
    Theta2_grad[:, 1:] = Theta2_grad[:, 1:] + (lambda_ / m) * Theta2[:, 1:]
    
    # Toodas las gradientes
    
    grad = np.concatenate([Theta1_grad.ravel(), Theta2_grad.ravel()])
    
    return J, grad


#### Pruebando el funcionamiento de la función de costo 

In [13]:
lambda_ = 0.1

J, _ = nnCostFunction(nn_params, input_layer_size, hidden_layer_size,  num_labels, X_norm, y_train, lambda_)
print("Costo en parámetros (Cargado): %.f " % J)
print('El costo debe esta cercano a          : 0.287629')

Costo en parámetros (Cargado): 13 
El costo debe esta cercano a          : 0.287629


In [14]:
z = np.array([-1, -0.5, 0, 0.5, 1])
g = sigmoidGradient(z)
print('Gradiente sigmoide evaluada con [-1 -0.5 0 0.5 1]:\n  ')
print(g)

Gradiente sigmoide evaluada con [-1 -0.5 0 0.5 1]:
  
[0.19661193 0.23500371 0.25       0.23500371 0.19661193]


#### . Inicializamso la función para inicializar los pesos 

La función randInitializeWeights se utiliza para inicializar aleatoriamente los pesos de una capa en una red neuronal. Esta inicialización aleatoria es crucial para evitar que todos los pesos comiencen con el mismo valor y se sincronicen durante el entrenamiento, lo que puede llevar a problemas de simetría y convergencia inadecuada durante el aprendizaje.

In [15]:
# Inicializamos los pesos de la red neuronal de forma aleatoria utilizando la función randInitializeWeights
def randInitializeWeights(L_in, L_out, epsilon_init = 0.12):
    
    """
        Inicializa aleatoriamente los pesos de una capa en una red neuronal.

        Parámetros
        ----------
        L_in:int
        Número de conexiones entrantes.

        L_salida: int
        Número de conexiones salientes.

        epsilon_init: flotante, opcional
        Rango de valores que puede tomar el peso de un uniforme
        distribución.

        Devoluciones
        -------
        W: tipo matriz
        El peso inicializado a valores aleatorios. Tenga en cuenta que W debería
        establecerse en una matriz de tamaño (L_out, 1 + L_in) como
        la primera columna de W maneja los términos de "sesgo".
    """
    
    """
    . Se crea una matriz W de dimensiones (L_out, 1 + L_in) inicializada con ceros.
    . L_out: representa el número de neuronas en la capa actual (conexiones salientes).
    . 1 + L_in: se refiere al número de entradas a cada neurona, incluyendo un término de "sesgo" (bias) que corresponde a la primera columna de la matriz W.
    """
    
    # Creamos una matriz con las dimensiones 
    W = np.zeros((L_out, 1 + L_in))
    W = np.random.rand(L_out, 1 + L_in) * 2 * epsilon_init - epsilon_init
    
    return W

In [16]:
# Llamamos a la funcion para inicializar los thetas aleatoriamente
print("Inicialización de parámetros de redes neuronales...")

initial_Theta1 = randInitializeWeights(input_layer_size, hidden_layer_size)
initial_Theta2 = randInitializeWeights(hidden_layer_size, num_labels)

# Desenrrollar parámetos
initial_nn_params = np.concatenate([initial_Theta1.ravel(), initial_Theta2.ravel()], axis=0)

Inicialización de parámetros de redes neuronales...


## **4. Entrenamiento del Modelo**

In [17]:
# Entrenamiento del modelo 

# Después de haber completado la tarea, cambie el maxiter a uno más grande
#? valor para ver cómo ayuda más formación.
# Especifica el número máximo de iteraciones que el algoritmo de optimización 
options = {'maxiter': 1000}

#? Probar con diferentes valores de lambda
lambda_ = 1

#? Cree una "taquigrafía" para minimizar la función de costos.
#? Ahora, costFunction es una función que toma solo un argumento.
costFunction = lambda p: nnCostFunction(p, input_layer_size, 
                                        hidden_layer_size, 
                                        num_labels, X_train, y_train, lambda_)

#? (los parámetros de la red neuronal)

res = optimize.minimize(costFunction,
                        initial_nn_params,
                        jac=True,
                        method='TNC',
                        options=options)

#? obtenemos la solución de la optimización
nn_params = res.x

#? Obtenemos Theta1 y Theta2 de nn_params
Theta1 = np.reshape(nn_params[:hidden_layer_size * (input_layer_size + 1)],
                    (hidden_layer_size, (input_layer_size + 1)))

Theta2 = np.reshape(nn_params[(hidden_layer_size * (input_layer_size + 1)):],
                    (num_labels, (hidden_layer_size + 1)))

C:\Users\gonza\AppData\Local\Temp\ipykernel_14460\1512527306.py:16: OptimizeWarning: Unknown solver options: maxiter
  res = optimize.minimize(costFunction,
C:\Users\gonza\AppData\Local\Temp\ipykernel_14460\2650450764.py:5: RuntimeWarning: overflow encountered in exp
  return 1.0 / (1.0 + np.exp(-z))


In [26]:
print(Theta1.shape)
print(Theta2.shape)

(10, 15)
(5, 11)


In [35]:
""" Mostramos las thetas """
print(Theta1)

[[-9.59355644e-02  9.59306968e-04 -2.35063860e-03 -1.46009302e-03
   6.64685960e-04  9.43760547e-04 -3.89394218e-04  2.95541677e-03
  -2.87475316e-03 -2.50237517e-03  2.04926724e-04 -2.97518664e-03
  -2.59384542e-03 -2.30145646e-03 -2.08926885e-03]
 [-1.02116005e-01 -1.89953180e-03  1.59501990e-03 -1.18460049e-03
  -2.68052541e-03 -2.94186337e-03 -3.02373407e-05 -8.83932818e-05
   5.56825559e-04  9.06687999e-04 -2.61788166e-03  4.48657834e-04
  -2.91987310e-03 -2.90827370e-03  2.80793433e-03]
 [ 4.34601542e+00 -1.39743421e-01 -4.01312062e-04 -8.68009965e-08
  -5.12361372e-03  1.26087984e-01  1.03073007e-02 -3.93071142e-01
  -1.28306501e-01  5.80095717e-01 -6.10488650e-06 -5.84609102e-05
  -1.21114897e-02 -5.67431440e-03 -5.17735410e-01]
 [ 6.67717650e-02 -9.76607471e-02 -3.73426687e-03 -5.74068612e-03
  -2.36898010e-02 -8.39171133e-03 -1.68392003e-02  5.65416885e-03
  -1.04361074e-03  1.19048551e-03  1.52529938e-01  1.29102946e-03
  -5.48409500e-02 -5.78258916e-02 -2.09343876e-03]
 [-8

In [36]:
""" Mostramos las thetas """
print(Theta2)

[[-1.85273331e-01  2.84010663e-03 -6.71177801e-04 -4.48170188e+00
   8.86235490e-01 -2.21524267e-01 -4.55419715e-01  1.11321564e+01
  -1.73024584e-01 -5.49892513e-02 -1.55643178e-01]
 [-2.80143358e-03 -2.91848173e-03  4.71346982e-03 -4.03392959e+00
  -4.10152388e-01  5.11074439e-02  2.58451504e-01 -7.82090846e+00
  -6.84386514e-02 -2.16451447e-01 -7.72439766e-02]
 [-5.84812709e-01 -2.79836282e-03 -2.20209854e-02  6.80848532e+00
   8.42536678e-02 -5.44540857e-01  1.96258648e-01 -9.85803165e+00
  -3.72930647e-01 -2.77976799e-01 -2.97284229e-01]
 [-6.32936876e-01 -2.59618493e-03 -7.68404157e-04 -2.08054432e+00
  -7.17944329e-01 -4.31904280e-01  6.51946564e-03 -5.13727291e+00
  -6.00683222e-01 -3.11069770e-01 -2.77553929e-01]
 [-6.07889648e-01  5.45152974e-04 -4.49375702e-02 -1.15958591e+01
  -8.37930805e-01 -2.59566943e-01  5.20413982e-02 -5.40753073e+00
  -3.73148845e-01 -7.70993108e-02 -6.08202488e-02]]


## **5. Evaluación del Modelo**

## 5.1 Predicciónes

Realizamos la evaluación del modelo esta vez con las predicciones con los datos de prueba del 20% aplicado para las pruebas de predicciones de X_test

In [28]:
# Inicializamos la función de predicción para predecir la etiqueta de una entrada dada una red neuronal entrenada
def predict(Theta1, Theta2, X):
    """
    Predecir la etiqueta de una entrada dada una red neuronal entrenada
    Genera la etiqueta predicha de X dados los pesos entrenados de un neural
    red (Theta1, Theta2)
    """
    
    #? Inicializamos los valores útiles
    m = X.shape[0]
    num_labels = Theta2.shape[0]
    
    #? Devolvemos las siguientes variables correctamente
    p = np.zeros(m)
    
    h1 = sigmoid(np.dot(np.concatenate([np.ones((m, 1)), X], axis=1), Theta1.T))
    h2 = sigmoid(np.dot(np.concatenate([np.ones((m, 1)), h1], axis=1), Theta2.T))
    p = np.argmax(h2, axis=1)
    
    return p

In [29]:
# Normalizamos los datos de prueba para aplicarlo en el modelo
X_test_norm, mu, sigma = featureNormalize(X_test)

# Mostramos los datos de X_test normalizados
X_test_norm

age  workclass    fnlwgt  education  education-num  occupation  \
0     0.781295  -1.967227  0.399354  -0.597642       0.353327   -0.850664   
1     2.629161   1.466918 -0.757945   1.208989      -0.031326   -0.612533   
2    -0.549169  -1.280398  0.084807  -0.339552       1.122631    1.054385   
3    -0.031766   2.153747  2.544566   1.208989      -0.031326   -0.612533   
4     0.116063   0.093260  0.117522   0.176628      -0.415978    1.292516   
...        ...        ...       ...        ...            ...         ...   
6417  0.263892   0.093260 -0.672883  -1.113822      -1.954587    1.768778   
6418 -0.031766  -2.654057 -0.773089  -1.113822      -1.954587   -1.565057   
6419  1.594356   0.093260 -0.442185   0.176628      -0.415978   -0.850664   
6420  0.855209   0.093260 -0.480183   1.208989      -0.031326   -0.850664   
6421  1.298697   0.093260 -0.362341  -1.113822      -1.954587    1.292516   

      relationship      race       sex  capital-gain  capital-loss  \
0        -0.895434  0.400085  0.704466     -0.146061     -0.221451   
1        -0.895434  0.400085  0.704466     -0.146061     -0.221451   
2        -0.895434 -1.934622  0.704466     -0.146061     -0.221451   
3         2.221695 -1.934622 -1.419515     -0.146061     -0.221451   
4         2.221695  0.400085 -1.419515     -0.146061     -0.221451   
...            ...       ...       ...           ...           ...   
6417     -0.895434  0.400085  0.704466     -0.146061     -0.221451   
6418      0.974843  0.400085 -1.419515     -0.146061     -0.221451   
6419     -0.895434  0.400085  0.704466     -0.146061     -0.221451   
6420     -0.272009  0.400085  0.704466      0.526314     -0.221451   
6421      1.598269  0.400085 -1.419515     -0.146061     -0.221451   

      hours-per-week  native-country     class  
0          -0.043445        0.293638 -0.567880  
1          -2.909118        0.293638 -0.567880  
2           0.752575        0.293638  1.760936  
3          -0.043445        0.165155 -0.567880  
4          -0.043445        0.293638  1.760936  
...              ...             ...       ...  
6417       -0.043445        0.293638 -0.567880  
6418       -2.272302        0.293638 -0.567880  
6419        0.752575        0.293638 -0.567880  
6420       -1.476282        0.293638 -0.567880  
6421       -0.043445        0.293638 -0.567880  

[6422 rows x 14 columns]

In [30]:
# Calculamos todas las predicciones con los datos de prueba  con los thetas obtenidos
pred = predict(Theta1, Theta2, X_test_norm)
print(pred)

[0 0 0 ... 0 0 2]


In [37]:
# Mostraremos todas las predicciones con los datos de prueba en una tabla concomparable con los datos reales
y_prediccion = pd.DataFrame({"Y-(Marital-Status)": y_test, "Y-(Predicción)": pred})

tabla_result = pd.concat([X_test, y_prediccion], axis=1)

tabla_result.head(50)

age  workclass  fnlwgt  education  education-num  occupation  \
0    49          1  233059          8             11           3   
1    74          6  109101         15             10           4   
2    31          2  199368          9             13          11   
3    38          7  462832         15             10           4   
4    40          4  202872         11              9          12   
5    60          0  134152          6              5           0   
6    43          4  221172          9             13           4   
7    27          4   59068         11              9           6   
8    53          5  220786         15             10          12   
9    45          4  156117          8             11          10   
10   40          4  252392          4              3           8   
11   66          0  129476          9             13           0   
12   43          4  136721          2              8           3   
13   26          4  122575          9             13           4   
14   36          4  201117          9             13           4   
15   18          4  217743          1              7          12   
16   27          4  183802         11              9           6   
17   31          4   33124         14             15          10   
18   40          2  141649          8             11          11   
19   43          6  101534         15             10           4   
20   46          4  165138          8             11           3   
21   31          4  110643          9             13           4   
22   32          4  159442         14             15          12   
23   27          4  159623         15             10          12   
24   20          4  137618         15             10          14   
25   54          4   35557         15             10          12   
26   40          7  175304         15             10          12   
27   53          4  153486         11              9          14   
28   46          4  117313          6              5           7   
29   45          4  128378         11              9           7   
30   31          4  132601         15             10           4   
31   18          0  344742          0              6           0   
32   34          4  159322          9             13           4   
33   41          4  206139         11              9          12   
34   36          7  212143          9             13           1   
35   31          4  215047          9             13           1   
36   52          4  145879         15             10          12   
37   38          4   27408         11              9           3   
38   20          4  250165         11              9          12   
39   53          6  105478          9             13           4   
40   42          6   32546         14             15          10   
41   27          7  271328          9             13          10   
42   62          4   65868         11              9          12   
43   90          2  214594          5              4          11   
44   65          2  103153          5              4           4   
45   35          4   22463         11              9           3   
46   25          0  100903          9             13           0   
47   18          4  183824          2              8           8   
48   45          4  285858          9             13          14   
49   64          4  316627          4              3           3   

    relationship  race  sex  capital-gain  capital-loss  hours-per-week  \
0              0     4    1             0             0              40   
1              0     4    1             0             0               4   
2              0     2    1             0             0              50   
3              5     2    0             0             0              40   
4              5     4    0             0             0              40   
5              1     2    1             0             0              35   
6         

Veridicamos cuantas clases predecidos tiene el conjunto de datos:

In [32]:
from collections import Counter

num_clases = len(np.unique(pred))
print("Número de clases:", num_clases)


# Conteo de datos por clase
conteo_clases = dict(Counter(pred))

# Imprimir el conteo de datos por clase
print("Clases podrían ser:")
for clase, conteo in conteo_clases.items():
    print(f"{clase} : {conteo} datos")

Número de clases: 2
Clases podrían ser:
0 : 4093 datos
2 : 2329 datos


## 5.2 Calcula la precisión (accuracy) 

Calculamos el error que se calcula como el número de predicciones incorrectas dividido por el número total de predicciones.
En este caso vamos a ir comparando los valores de Y predichos con los datos de y_train (elementos principales) y calcular la precisión:

In [34]:
# Calculamos la prección del modelo 
def accuracy(y_pred, y):
    return np.sum(y_pred == y) / len(y)


precision_reg = accuracy(pred, y_test)

print(f"La precisión del modelo es: {precision_reg * 100:.2f}%")

La precisión del modelo es: 61.52%


# **6. Conclusión**

Segun los resultados tanto de predicciónes y la precisión del entrenamiento del modelo de red neuronal es de 61.52% y podemos ver que
el modelo está rindiendo bien, en este caso que es el objetivo de predecir si un hombre 
adulto se identifica a una clase de 0 al 4. La precisión del 61.52% es un buen resultado, y podemos ver que la probabilidad de que un hombre adulto se identifique a una clase de 0 al 4 es del 61.52%.
